query-node/mappings/content-directory/transaction.ts

```diff
@@ -81,33 +81,45 @@ async function getNextEntityId(db: DB): Promise<number> {
   return e.nextId
 }
 
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export async function contentDirectory_TransactionFailed(db: DB, event: SubstrateEvent): Promise<void> {
+  debug(`TransactionFailed event: ${JSON.stringify(event)}`)
+
+  const failedOperationIndex = event.params[1].value as number
+  const operations = decode.getOperations(event)
+
+  if (operations.length === 0 || operations.length === 1) return
+
+  const successfulOperations = operations.filter((op, index) => index < failedOperationIndex)
+
+  const {
```

Comment:
You can factor out this common code in `contentDirectory_TransactionFailed` and `contentDirectory_TransactionCompleted` into a function `applyOperations(operations: IBatchOperation)`

Reply:
Done.